### ISSUE TITLE
.eslintignore File Location and related Search Pattern

### ISSUE THREAD

**Issue by octotree** [eslint/eslint#8090]
**TL;DR:** Is there a special reason, why the .eslintignore file is not searched
relative to the files which are linted?

Two weeks ago, I started using eslint with syntactic within vim. For the first few days, everything worked great, but then I came across that lib which had a totally different style than my code. So I thought I could simply create a `.eslintignore` file. So I did and placed it in the same directory as my `.eslintrc`. But it wasn't found. So I placed it next to the file I wanted to edit, but it wasn't found either. So I took a look at the documentation and found that eslint was searching in the current work directory.

I found this pretty weird because normally my editor does not care from which directory I edit a file. So I went to your gitter and had a short conversation with @platinumazure. I went on and wrote a little patch which fixes the issue (for me) and should break as little as possible.

The patch (https://paste.pound-python.org/show/w5HghbQWb0HjHfMxPyJc/) solves the problem by first looking in the CWD as before and extends its search for a `.eslintignore` file starting with the directory of the first target js-file (argument from cli), upwards in the directory tree. If it cannot find an `.eslintignore` file it continues relative to the next target file. The search terminates as soon as a `.eslintignore` file is found or all parent directories have been searched. As directory trees are normally not that deep (who uses more than 20 directories in a path?), I think the overhead should be okay. I do not know with how many files as arguments people use eslint but when they would use many files directly as cli arguments (200?) in a deeply nested directory on a hard drive with cold FS-cache, it could take a few seconds to probe all those directories, but I think this is more of an edge case.

So what does the patch break: As the patch still searches the CWD first, the breakage should be minimal. The only case I am aware of is when someone uses a CWD without `.eslintignore` but has some `.eslintignore` file in a parent directory with matching paths and does not want to ignore those files. Very unlikely, but possible.

When I presented the patch to @platinumazure he advised me to either add tests and create a PR or create an issue first (to spare my resources), to ask, if there is a special reason, why this functionality was not built into eslint before. Since I did not have the time to create tests in the last two weeks I decided to start with the issue.

**So what do you think?**

* **ESLint Version:** v3.15.0
* **Node Version:** v7.5.0
* **npm Version:** 4.1.2

**Comment by not-an-aardvark** [comment]
I think this is a duplicate of https://github.com/eslint/eslint/issues/6759. It will be fixed by https://github.com/eslint/eslint/pull/7678 in the next major release.

**Comment by platinumazure** [comment]
@not-an-aardvark No, this is not the same thing.

#7678 focuses on how to interpret patterns in an .eslintignore file, once that file is found. Prior to #7678, they're interpreted relative to CWD; after #7678, they're interpreted relative to the location of the .eslintignore file. This fixes cases where someone is using `--ignore-path` to point to an ignore file that's not in CWD.

This issue is about locating the .eslintignore file at all. Currently, ESLint only looks in the current working directory for .eslintignore. Now, we can't support a *lot* of complexity around looking for this file (because it must be done prior to *any* traversal for lintable files), but I do think we could maybe support something where we iterate up ancestor directories from CWD until we find an .eslintignore file (or possibly have some other stopping point, such as a directory containing an .eslintrc with `"root": true`), and then use that as the ignore file. (This somewhat depends on #7678 being merged, in order to behave predictably when running ESLint from subdirectories.)

The use case which is motivating @octotree is for editor integrations, which might open a file in a descendant CWD and which might invoke ESLint from that directory rather than traversing up to find an `.eslintrc` with `"root": true`.

@octotree Let me know if I've missed or misstated anything. @not-an-aardvark Let me know if I can explain this better.

**Comment by octotree** [comment]
@platinumazure very well clarified :+1:

**Comment by IanVS** [comment]
Personally, I hope that this behavior is _not_ changed, because the presence of a `.eslintignore` file is the only reliable indicator I've found of determining the project root.  There are several tools in the ecosystem which depend on eslint being run from project root (for example the webpack resolver for `eslint-plugin-import`).  So, within Atom's `linter-eslint`, we check to see if there is a `.eslintignore` file, and if we can find one, we `chdir` to that directory.  If eslint begins to allow `.eslintignore` files from other locations, we won't be able to rely on that behavior.

**Comment by octotree** [comment]
@IanVS I see your point but actually `.eslintignore` files are searched for in the CWD (which is the users home directory by default). The project root is defined by `"root": true` within a configuration file (e.g. .eslintrc) as documented [here](http://eslint.org/docs/user-guide/configuring#configuration-cascading-and-hierarchy).

Funny enough, the documentation also states, that the `.eslintignore` file should be created within your project's root, but I didn't find any code indicating that behavior. I mean, this issue is about the fact, that eslint is not searching the `.eslintignore` file in the project's root as documented, but in the current work directory instead.

Had you trouble searching for `"root": true` or was there another reason why you do not rely on that option?

**Comment by IanVS** [comment]
@octotree I believe that much of the code also assumes that you will be running eslint from the project root.  When used as a CLI tool, that's certainly usually the case.  But as you've found out, it's a bit of a challenge for other tools like text editors.  

The option `root: true` only really stops configuration merging.  When eslint finds a config with that option, it will not merge any other configurations in ancestor directories.  The use cases for it are varied, and do not always indicate the project root.  I've most commonly seen it used within text fixture directories.  Many projects do not use it at all.

Note that until #7678 lands, it will not be possible to use a `.eslintignore` file that is not in CWD.  I can't help but think there are going to be other challenges in supporting running ESLint from a directory that is not the project root (like the webpack resolver I mentioned earlier).

**Comment by octotree** [comment]
@IanVS I am well aware of the #7678 dependency here.

So the problem I see here is that on the one hand eslint seem to depend heavily on the project root while at the same time there is no clear definition where to find it. Some say it is the directory with the `.eslintignore` file, others say it is the directory with the `.eslinrc` containing `"root": true`. I do not really care which one is the truth/better.

The only thing I do NOT believe to be meaningful is to assume that the CWD is the project's root.

Just to give an example. In some of my projects I have multiple "sub-projects" in close proximity:

- /app (Browser App)
- /cordova (WebView App)
- /aws/lambda (JS Backend)
- /playground/... (various small demos)

Most of the time you are just in one of those sub-projects but sometimes you want to open multiple files from different sub-projects at the same time or you just want to edit a file from another sub-project for a short time.

Always closing the editor and starting it from another subdirectory sucks, but even when you run eslint from the command line (e.g. with --fix) it is really annoying if you always have to change to the correct directory first.

On the other hand, if we would try to find the project's root based on the file path which eslint has to lint, it would make much more sense and if I am not mistaken eslint does so already when it searches for the `.eslintrc`.

So an alternative to the initial patch could be, to agree on a definition of the project root which is based either on `.eslinrc` XOR `.eslintignore` (ignoring the CWD), let eslint find the project root at the beginning and change into the project`s root directory before it starts the lint-process?

**Comment by octotree** [comment]
@nzakas actually, it seems there was enough interest as someone implemented it. Current documentation states:

> ESLint will automatically look for them in the directory of the file to be linted, and in successive parent directories all the way up to the root directory of the filesystem (unless root: true is specified).

After the negative feedback here I just went on and forgot about the issue. Some day I noticed that ESLint behaves better now, but didn't remember this issue was still open. So thanks for closing it and thanks to the person who fixed the behavior :+1:
